patches/tModLoader/Terraria/ModLoader/ContentCache.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Terraria.ModLoader;
+
+/// <summary>
+/// Container object for <see cref="ILoadable"/> instances added for a <see cref="Mod"/>
+/// </summary>
+internal class ContentCache
+{
+	private class Cache<T> where T : ILoadable {
+		private static Dictionary<string, Cache<T>> _instancesPerMod;
+
+		public static Cache<T> GetOrCreate(ContentCache cache) {
+			if (_instancesPerMod is null) {
+				_instancesPerMod = new();
+				OnUnload += Unload;
+			}
+
+			string key = cache._mod.Name;
+
+			if (!_instancesPerMod.TryGetValue(key, out Cache<T> instance))
+				_instancesPerMod[key] = instance = new Cache<T>(cache);
+
+			return instance;
+		}
+
+		private static void Unload() {
+			_instancesPerMod = null;
+		}
+
+		private readonly ContentCache _source;
+		private IReadOnlyList<T> _cache;
+
+		private Cache(ContentCache source) {
+			_source = source;
+		}
+
+		public IEnumerable<T> GetOrCacheContent() {
+			IEnumerable<T> lazyContent = _source._content.OfType<T>();
+
+			if (_source._mod.loading) {
```

Comment:
This is good.

After applying my above advice, please also add a bail-out for when there's no content at all because the mod hasn't yet been loaded.